File: ethical_ontology/chaincode/dao_contract.go
package reputation

type Proposal struct {
	ID           string
	Description  string
	VotesFor     float64
	VotesAgainst float64
	Voters       map[string]bool // To prevent double voting
	Active       bool
}

type DAOContract struct {
	proposals  map[string]*Proposal
	reputation *ReputationContract
	quorum     float64
}

func NewDAOContract(repContract *ReputationContract, quorum float64) *DAOContract {
	return &DAOContract{
		proposals:  make(map[string]*Proposal),
		reputation: repContract,
		quorum:     quorum,
	}
}

func (d *DAOContract) ProposeRule(id string, description string, proposerID string) bool {
	if _, exists := d.proposals[id]; exists {
		return false
	}
	// Check proposer reputation
	if d.reputation.GetReputation(proposerID) < 30 {
		return false
	}
	d.proposals[id] = &Proposal{
		ID:           id,
		Description:  description,
		VotesFor:     0,
		VotesAgainst: 0,
		Voters:       make(map[string]bool),
		Active:       true,
	}
	return true
}

func (d *DAOContract) Vote(proposalID string, agentID string, voteFor bool, weight int) bool {
	prop, exists := d.proposals[proposalID]
	if !exists || !prop.Active {
		return false
	}
	if prop.Voters[agentID] {
		return false
	}
	voteWeight := d.reputation.QuadraticVote(agentID, weight)
	if voteFor {
		prop.VotesFor += voteWeight
	} else {
		prop.VotesAgainst += voteWeight
	}
	prop.Voters[agentID] = true
	return true
}

func (d *DAOContract) Enact(proposalID string) bool {
	prop, exists := d.proposals[proposalID]
	if !exists || !prop.Active {
		return false
	}
	totalVotes := prop.VotesFor + prop.VotesAgainst
	if totalVotes == 0 {
		return false
	}
	if prop.VotesFor/totalVotes >= d.quorum {
		prop.Active = false
		// Update chaincode or ethical rules here
		return true
	}
	return false
}

func (d *DAOContract) GetProposal(id string) *Proposal {
	return d.proposals[id]
}

func (d *DAOContract) GetAllProposals() []*Proposal {
	var all []*Proposal
	for _, p := range d.proposals {
		all = append(all, p)
	}
	return all
}

File: ethical_ontology/chaincode/dao_contract.py
"""
import logging
from typing import Dict, Any, List
from .base_contract import BaseSmartContract
from .virtue_reputation import VirtueReputationContract

logger = logging.getLogger(__name__)

class Proposal:
    def __init__(self, id: str, description: str):
        self.id = id
        self.description = description
        self.votes_for = 0.0
        self.votes_against = 0.0
        self.voters = set()  # To prevent double voting
        self.active = True

class DAOContract(BaseSmartContract):
    """
    Smart contract for DAO governance of ethical rules.
    
    Allows stakeholders to propose, vote on, and enact ethical rule changes
    using reputation-weighted voting.
    """
    
    def __init__(self, reputation_contract: VirtueReputationContract, quorum: float = 0.51):
        super().__init__("DAOContract", "1.0.0")
        self.proposals: Dict[str, Proposal] = {}
        self.reputation = reputation_contract
        self.quorum = quorum
        self.total_voting_power = 0.0  # Updated dynamically
        logger.info(f"Initialized DAOContract with quorum {quorum}")
    
    def propose_rule(self, proposal_id: str, description: str, proposer_id: str) -> bool:
        """Propose a new ethical rule change."""
        if proposal_id in self.proposals:
            logger.warning(f"Proposal {proposal_id} already exists")
            return False
        
        # Check if proposer has sufficient reputation
        rep = self.reputation.get_agent_reputation(proposer_id)
        if not rep or rep['overall_reputation'] < 0.3:
            logger.warning(f"Proposer {proposer_id} has insufficient reputation")
            return False
        
        self.proposals[proposal_id] = Proposal(proposal_id, description)
        self._log_call("propose_rule", {"id": proposal_id, "desc": description[:50]}, True)
        return True
    
    def vote(self, proposal_id: str, agent_id: str, vote_for: bool) -> bool:
        """Cast a reputation-weighted vote on a proposal."""
        if proposal_id not in self.proposals:
            return False
        proposal = self.proposals[proposal_id]
        if not proposal.active:
            return False
        if agent_id in proposal.voters:
            logger.warning(f"Agent {agent_id} already voted on {proposal_id}")
            return False
        
        # Get agent's reputation score
        rep = self.reputation.get_agent_reputation(agent_id)
        if not rep:
            return False
        weight = rep['overall_reputation']
        
        if vote_for:
            proposal.votes_for += weight
        else:
            proposal.votes_against += weight
        
        proposal.voters.add(agent_id)
        self._log_call("vote", {"proposal": proposal_id, "agent": agent_id, "for": vote_for}, True)
        return True
    
    def enact(self, proposal_id: str) -> bool:
        """Enact a proposal if quorum is met."""
        if proposal_id not in self.proposals:
            return False
        proposal = self.proposals[proposal_id]
        if not proposal.active:
            return False
        
        total_votes = proposal.votes_for + proposal.votes_against
        if total_votes == 0:
            return False
        
        approval_ratio = proposal.votes_for / total_votes
        if approval_ratio >= self.quorum:
            proposal.active = False
            # Here: Tie to other chains - update ethical rules
            # For simplicity, log and update state
            self.update_state(f"enacted_{proposal_id}", proposal.description)
            self._log_call("enact", {"proposal": proposal_id}, True)
            
            # TODO: Integrate with other contracts to apply rule change
            logger.info(f"Proposal {proposal_id} enacted with {approval_ratio:.2f} approval")
            return True
        
        logger.info(f"Proposal {proposal_id} failed enactment: {approval_ratio:.2f} < {self.quorum}")
        return False
    
    def get_proposal(self, proposal_id: str) -> Dict[str, Any]:
        """Get details of a specific proposal."""
        if proposal_id not in self.proposals:
            return {}
        p = self.proposals[proposal_id]
        return {
            "id": p.id,
            "description": p.description,
            "votes_for": p.votes_for,
            "votes_against": p.votes_against,
            "active": p.active,
            "approval_ratio": p.votes_for / (p.votes_for + p.votes_against) if (p.votes_for + p.votes_against) > 0 else 0.0
        }
    
    def get_all_proposals(self) -> List[Dict[str, Any]]:
        """Get all current proposals."""
        return [self.get_proposal(pid) for pid in self.proposals.keys()]
    
    # Abstract methods implementation
    def check_compliance(self, action_description: str, **kwargs) -> Dict[str, Any]:
        """For DAO, compliance could check if action aligns with enacted rules."""
        # For now, stub
        return {"compliant": True, "confidence": 1.0, "reasoning": "DAO compliance not implemented", "rule_applied": "stub"}
    
    def get_applicable_rules(self) -> List[Dict[str, Any]]:
        """Return DAO governance rules."""
        return [
            {"rule_id": "dao_001", "rule_name": "Proposal Creation", "description": "Rules for creating proposals", "parameters": ["proposal_id", "description", "proposer_id"]},
            {"rule_id": "dao_002", "rule_name": "Voting", "description": "Reputation-weighted voting", "parameters": ["proposal_id", "agent_id", "vote_for"]},
            {"rule_id": "dao_003", "rule_name": "Enactment", "description": "Quorum-based enactment", "parameters": ["proposal_id"]}
        ]

""" 